app/javascript/mastodon/components/dropdown_menu.tsx

```diff
@@ -0,0 +1,532 @@
+import {
+  useState,
+  useEffect,
+  useRef,
+  useCallback,
+  cloneElement,
+  Children,
+} from 'react';
+
+import classNames from 'classnames';
+import { Link } from 'react-router-dom';
+
+import type { Map as ImmutableMap } from 'immutable';
+
+import Overlay from 'react-overlays/Overlay';
+import type {
+  OffsetValue,
+  UsePopperOptions,
+} from 'react-overlays/esm/usePopper';
+
+import { fetchRelationships } from 'mastodon/actions/accounts';
+import {
+  openDropdownMenu,
+  closeDropdownMenu,
+} from 'mastodon/actions/dropdown_menu';
+import { openModal, closeModal } from 'mastodon/actions/modal';
+import { CircularProgress } from 'mastodon/components/circular_progress';
+import { isUserTouching } from 'mastodon/is_mobile';
+import type {
+  MenuItem,
+  ActionMenuItem,
+  ExternalLinkMenuItem,
+} from 'mastodon/models/dropdown_menu';
+import { useAppDispatch, useAppSelector } from 'mastodon/store';
+
+import type { IconProp } from './icon';
+import { IconButton } from './icon_button';
+
+let id = 0;
+
+const isMenuItem = (item: unknown): item is MenuItem => {
+  if (item === null) {
+    return true;
+  }
+
+  return !!(item as MenuItem)?.text;
+};
+
+const isActionItem = (item: unknown): item is ActionMenuItem => {
+  if (!item) {
+    return false;
+  }
+
+  return typeof (item as ActionMenuItem).action === 'function';
+};
+
+const isExternalLinkItem = (item: unknown): item is ExternalLinkMenuItem => {
+  if (!item) {
+    return false;
+  }
+
+  return !!(item as ExternalLinkMenuItem).href;
+};
+
+type RenderItemFn<Item = MenuItem> = (
+  arg0: Item,
+  arg1: number,
+  arg2: {
+    onClick: (e: React.MouseEvent) => void;
+    onKeyUp: (e: React.KeyboardEvent) => void;
+  },
+) => React.ReactNode;
+
+type RenderHeaderFn<Item = MenuItem> = (arg0: Item[]) => React.ReactNode;
+
+interface DropdownMenuProps<Item = MenuItem> {
+  items?: Item[];
+  loading?: boolean;
+  scrollable?: boolean;
+  onClose: () => void;
+  openedViaKeyboard: boolean;
+  renderItem?: RenderItemFn<Item>;
+  renderHeader?: RenderHeaderFn<Item>;
+  onItemClick: (e: React.MouseEvent | React.KeyboardEvent) => void;
+}
+
+const DropdownMenu = <Item = MenuItem,>({
```

Comment:
i notice GH and IDE formatting breaks with this comma which is added by Prettier. i did some poking but wasn't able to resolve this issue annoyingly.